### PR TITLE
feat: add jwk generator with docker support

### PR DESCRIPTION
**Pull request by ealogar**

### CODE REVIEW

File: delivery/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ENV MAVEN_CONFIG "/root/.m2"
+ENV MAVEN_OPTS "-Dmaven.wagon.http.ssl.insecure=true -Dmaven.wagon.http.ssl.allowall=true"
+
+ARG AWS_SECRET_ACCESS_KEY
+ARG AWS_ACCESS_KEY_ID
+
+WORKDIR /opt/app
+
+COPY pom.xml ./pom.xml
+COPY src/ ./src/
```
Comment (mmolinos) [inline delivery/docker/Dockerfile]:
```suggestion
COPY pom.xml ./pom.xml

RUN mvn dependency:resolve-plugins dependency:resolve clean package

COPY src/ ./src/
``` 
maybe this change for improving the compilation time when you don't change the pom?

File: delivery/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+RUN mvn clean package
+
+FROM openjdk:11
```
Comment (mmolinos) [inline delivery/docker/Dockerfile]:
```suggestion
FROM openjdk:11.0.6-slim
```

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            <version>[1.64,)</version>
+        </dependency>
+        <dependency>
+            <groupId>commons-cli</groupId>
+            <artifactId>commons-cli</artifactId>
+            <version>1.2</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
```
Comment (mmolinos) [inline pom.xml:37]: maybe it's better to use Jackson?? Think that it has a better performance

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+import java.security.Security;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public class Entrypoint {
+  private static String DEFAULT_KEY_SIZE = "2048";
+  private static Options options;
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
  private static final String DEFAULT_KEY_SIZE = "2048";
  private static final Options OPTIONS = new Options();
  private static final JsonElement JSON = new JsonParser();
  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+  public static void main(String[] args) {
+
+    Security.addProvider(new BouncyCastleProvider());
+
+    options = new Options();
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.List;
+import java.util.Objects;
+
+public class Entrypoint {
+  private static String DEFAULT_KEY_SIZE = "2048";
+  private static Options options;
+
+  public static void main(String[] args) {
+
+    Security.addProvider(new BouncyCastleProvider());
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java:39]: maybe add a comment for improving who is this and why is this mandatory?

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      if (keySize % 8 != 0) {
+        printUsageAndExit("Key size (in bits) must be divisible by 8, got " + keySize);
+      }
+
+      JWK jwk = RSAKeyMaker.make(keySize, keyUse, algorithm);
+
+      // round trip it through GSON to get a prettyprinter
+      Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+      printKey(keySet, jwk, gson);
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]: extract this to a new method for clear the objective of this action

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+
+  private static void printKey(boolean keySet, JWK jwk, Gson gson) {
+    if (keySet) {
+      JWKSet jwkSet = new JWKSet(jwk);
+      JsonElement json = new JsonParser().parse(jwkSet.toJSONObject(false).toJSONString());
+      System.out.println(gson.toJson(json));
+    } else {
+      JsonElement json = new JsonParser().parse(jwk.toJSONString());
+      System.out.println(gson.toJson(json));
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
      System.out.println(GSON.toJson(json));
```
but is it not better to create a new file with the new keys?

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  // print out a usage message and quit
+  private static void printUsageAndExit(String message) {
+    if (message != null) {
+      System.err.println(message);
+    }
+
+    List<String> optionOrder = Arrays.asList("s", "u");
+
+    HelpFormatter formatter = new HelpFormatter();
+    //formatter.setOptionComparator((o1, o2) -> optionOrder.indexOf(((Option) o1).getOpt()) - optionOrder.indexOf(((Option) o2).getOpt()));
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    } catch (ParseException e) {
+      printUsageAndExit("Failed to parse arguments: " + e.getMessage());
+    }
+  }
+
+
+  private static void printKey(boolean keySet, JWK jwk, Gson gson) {
+    if (keySet) {
+      JWKSet jwkSet = new JWKSet(jwk);
+      JsonElement json = new JsonParser().parse(jwkSet.toJSONObject(false).toJSONString());
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
      JsonElement json = JSON.parse(jwkSet.toJSONObject(false).toJSONString());
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+      System.out.println(gson.toJson(json));
+    } else {
+      JsonElement json = new JsonParser().parse(jwk.toJSONString());
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
      JsonElement json = JSON.parse(jwk.toJSONString());
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+      Integer keySize = Integer.decode(size);
+      if (keySize % 8 != 0) {
+        printUsageAndExit("Key size (in bits) must be divisible by 8, got " + keySize);
+      }
+
+      JWK jwk = RSAKeyMaker.make(keySize, keyUse, algorithm);
+
+      // round trip it through GSON to get a prettyprinter
+      Gson gson = new GsonBuilder().setPrettyPrinting().create();
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java (hunk 7 above)
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
      printKey(keySet, jwk);
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+      printKey(keySet, jwk, gson);
+    } catch (NumberFormatException e) {
+      printUsageAndExit("Invalid key size: " + e.getMessage());
+    } catch (ParseException e) {
+      printUsageAndExit("Failed to parse arguments: " + e.getMessage());
+    }
+  }
+
+
+  private static void printKey(boolean keySet, JWK jwk, Gson gson) {
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
  private static void printKey(boolean keySet, JWK jwk) {
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    if (keySet) {
+      JWKSet jwkSet = new JWKSet(jwk);
+      JsonElement json = new JsonParser().parse(jwkSet.toJSONObject(false).toJSONString());
+      System.out.println(gson.toJson(json));
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java]:
```suggestion
      System.out.println(GSON.toJson(json));
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * Generate a new RSAKey with the kid specified.
+   *
+   * @param keySize
+   * @param keyUse
+   * @param keyAlg
+   * @return
+   */
+  public static RSAKey make(Integer keySize, KeyUse keyUse, Algorithm keyAlg) {
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java:40]:
```suggestion
  public static RSAKey createRsaKey(Integer keySize, KeyUse keyUse, Algorithm keyAlg) {
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+public class RSAKeyMaker {
+
+  /**
+   * Generate a kid using the sha-1 digest of public exponent
+   *
+   * @param publicModulus
+   * @return
+   * @throws NoSuchAlgorithmException
+   */
+  public static String makeKid(BigInteger publicModulus) throws NoSuchAlgorithmException {
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java:24]:
```suggestion
  public static String createKid(BigInteger publicModulus) throws NoSuchAlgorithmException {
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+   * @return
+   */
+  public static RSAKey make(Integer keySize, KeyUse keyUse, Algorithm keyAlg) {
+
+    try {
+      KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+      generator.initialize(keySize);
+      KeyPair kp = generator.generateKeyPair();
+
+      RSAPublicKey pub = (RSAPublicKey) kp.getPublic();
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java]:
```suggestion
      RSAPublicKey publicKey = (RSAPublicKey) kp.getPublic();
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+      RSAPrivateKey priv = (RSAPrivateKey) kp.getPrivate();
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java]:
```suggestion
      RSAPrivateKey privateKey = (RSAPrivateKey) kp.getPrivate();
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+   * @param keyUse
+   * @param keyAlg
+   * @return
+   */
+  public static RSAKey make(Integer keySize, KeyUse keyUse, Algorithm keyAlg) {
+
+    try {
+      KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+      generator.initialize(keySize);
+      KeyPair kp = generator.generateKeyPair();
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java]:
```suggestion
      KeyPair keyPair = generator.generateKeyPair();
```

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+      RSAKey rsaKey = new RSAKey.Builder(pub)
+        .privateKey(priv)
+        .keyUse(keyUse)
+        .algorithm(keyAlg)
+        .keyID(kid)
+        .build();
+
+      return rsaKey;
+    } catch (NoSuchAlgorithmException e) {
+      return null;
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/RSAKeyMaker.java:61]: maybe print an error message?

File: delivery/docker/Dockerfile (hunk 1 above)
Comment (ealogar) [reply delivery/docker/Dockerfile]: ok, but notice that I will run this dockerfile once in life...

File: src/main/java/com/telefonica/jwk/Entrypoint.java (hunk 8 above)
Comment (ealogar) [reply src/main/java/com/telefonica/jwk/Entrypoint.java]: We want to parse the output, support for file may be used in future ...

File: src/main/java/com/telefonica/jwk/RSAKeyMaker.java (hunk 16 above)
Comment (ealogar) [reply src/main/java/com/telefonica/jwk/RSAKeyMaker.java:24]: I like more makeKid ... 💃

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 21]
```diff
@@ -27,28 +26,52 @@
 public class Entrypoint {
   private static final String DEFAULT_KEY_SIZE = "2048";
   private static final Options OPTIONS = new Options();
-  private static final JsonElement JSON = new JsonParser();
+  private static final JsonParser JSON = new JsonParser();
   private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static String size = null;
+  private static KeyUse keyUse = null;
+  private static Algorithm algorithm = JWEAlgorithm.RSA_OAEP_256;
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java:33]:
```suggestion
  private static final Algorithm algorithm = JWEAlgorithm.RSA_OAEP_256;
```

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 22]
```diff
@@ -27,28 +26,52 @@
 public class Entrypoint {
   private static final String DEFAULT_KEY_SIZE = "2048";
   private static final Options OPTIONS = new Options();
-  private static final JsonElement JSON = new JsonParser();
+  private static final JsonParser JSON = new JsonParser();
   private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static String size = null;
+  private static KeyUse keyUse = null;
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java:32]: I think that was better as local variables

File: src/main/java/com/telefonica/jwk/Entrypoint.java [hunk 23]
```diff
@@ -27,28 +26,52 @@
...
+  private static Algorithm algorithm = JWEAlgorithm.RSA_OAEP_256;
+  private static boolean keySet = true;
```
Comment (mmolinos) [inline src/main/java/com/telefonica/jwk/Entrypoint.java:34]: same with this. It was better as local var

**Review by mmolinos — APPROVED** [review]

File: src/main/java/com/telefonica/jwk/Entrypoint.java (hunk 21 above)
Comment (ealogar) [reply src/main/java/com/telefonica/jwk/Entrypoint.java:33]: No puede ser final porque la modificamos después

File: src/main/java/com/telefonica/jwk/Entrypoint.java (hunk 23 above)
Comment (ealogar) [reply src/main/java/com/telefonica/jwk/Entrypoint.java:34]: entonces tendría que meter un objeto para obtener el resultado del parseo ...

**Pull request merged**